### DISCUSSION TITLE
Handling of CellID in scv.utils.merge

### DISCUSSION THREAD

**Discussion by ifanirene** [theislab/scvelo#524] [Q&A] [answered]
Dear scvelo team,

I'm trying to perform add loom files to pre-processed adata that is consisted of three samples (annotated with -1,-2,-3 in cellID). I saw in solved issue (#36) you mentioned that the merge module would first clean up the CellID before merging. If this is the case how would the sample correspondence be retained when merging the data?

Would you recommend the following workflow?
1. split adata by sample -> adata1, adata2, adata3
2. adata1=scv.utils.merge(adata1,ldata1)
    adata2=scv.utils.merge(adata2,ldata2)
    adata3=scv.utils.merge(adata3,ldata3)
3. adata = adata1.concatenate(adata2, adata3)

Thanks,
Irene

**Comment by WeilerP** [comment, marked as answer]
@ifanirene, I suppose your `ldataX` objects are the AnnData objects you get from the loom files?

`scv.utils.merge` only "cleans" the observation names if there are no overlap between `adata.obs_names` and `ldata.obs_names`:

https://github.com/theislab/scvelo/blob/1659cc8e00a45fcf87cd80a7013aae5531744613/scvelo/read_load.py#L165
https://github.com/theislab/scvelo/blob/1659cc8e00a45fcf87cd80a7013aae5531744613/scvelo/read_load.py#L168-L171

`scv.utils.clean_obs_names` removes unwanted parts from the observation names and saves the removed part in `adata.obs['sample_batch']`.

https://github.com/theislab/scvelo/blob/1659cc8e00a45fcf87cd80a7013aae5531744613/scvelo/read_load.py#L126-L131

You can use your workflow. Alternatively, you can also first concatenate the `ldataX` and then merge them with `adata`. Should be equivalent, though.

**Reply by ifanirene** [reply]
@WeilerP Thank you for the explanation! This makes a lot of sense. Yes the ldata are AnnData built from loom files. The sample annotation in the merged ldata and adata was slightly different so they got removed. I know how to fix it now.

Cheers,
Irene
